dask/dataframe/tests/test_dataframe.py

```diff
@@ -2926,6 +2926,23 @@ def test_astype():
     assert_eq(a.x.astype(float), df.x.astype(float))
 
 
+def test_astype_errors():
+    df = pd.DataFrame(
+        {"x": [1, 2, 3, "invalid_string"], "y": [10, 20, 30, 40]},
+        index=[10, 20, 30, 40],
+    )
+    a = dd.from_pandas(df, 2)
+
+    msg = "could not convert string to float: 'invalid_string'"
+    with pytest.raises(ValueError, match=msg):
+        assert_eq(a.astype(float), df.astype(float))
+    with pytest.raises(ValueError, match=msg):
+        assert_eq(a.x.astype(float), df.x.astype(float))
+
+    assert_eq(a.astype(str, errors="ignore"), df.astype(str, errors="ignore"))
+    assert_eq(a.x.astype(str, errors="ignore"), df.x.astype(str, errors="ignore"))
```

Comment:
I'm having some trouble understanding what's going on here. Setting the new type to `float` causes an error. I suspect this is due to using `numeric_equal` but I'm not sure how the metadata should come into play here. 

Any thoughts?

```
    def test_astype_errors():
        df = pd.DataFrame(
            {"x": [1, 2, 3, "invalid_string"], "y": [10, 20, 30, 40]},
            index=[10, 20, 30, 40],
        )
        a = dd.from_pandas(df, 2)
    
        msg = "could not convert string to float: 'invalid_string'"
        with pytest.raises(ValueError, match=msg):
            assert_eq(a.astype(float), df.astype(float))
        with pytest.raises(ValueError, match=msg):
            assert_eq(a.x.astype(float), df.x.astype(float))
    
>       assert_eq(a.astype(float, errors="ignore"), df.astype(float, errors="ignore"))

dask/dataframe/tests/test_dataframe.py:2942: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dask/dataframe/utils.py:795: in assert_eq
    a = _check_dask(a, check_names=check_names, check_dtypes=check_dtypes)
dask/dataframe/utils.py:741: in _check_dask
    assert_dask_dtypes(dsk, result)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

ddf = Dask DataFrame Structure:
                     x        y
npartitions=2                  
10             float64  float64
30                 ...      ...
40                 ...      ...
Dask Name: astype, 4 tasks
res =                  x     y
10               1  10.0
20               2  20.0
30               3  30.0
40  invalid_string  40.0, numeric_equal = True

    def assert_dask_dtypes(ddf, res, numeric_equal=True):
        """Check that the dask metadata matches the result.
    
        If `numeric_equal`, integer and floating dtypes compare equal. This is
        useful due to the implicit conversion of integer to floating upon
        encountering missingness, which is hard to infer statically."""
    
        eq_type_sets = [{"O", "S", "U", "a"}]  # treat object and strings alike
        if numeric_equal:
            eq_type_sets.append({"i", "f", "u"})
    
        def eq_dtypes(a, b):
            return any(
                a.kind in eq_types and b.kind in eq_types for eq_types in eq_type_sets
            ) or (a == b)
    
        if not is_dask_collection(res) and is_dataframe_like(res):
            for col, a, b in pd.concat([ddf._meta.dtypes, res.dtypes], axis=1).itertuples():
>               assert eq_dtypes(a, b)
E               AssertionError

dask/dataframe/utils.py:894: AssertionError
```

Reply:
With `.astype(float, errrors="ignore")`, we don't actually know the dtype of the output until the graph is actually run. The dtype now depends on the values of the data. That means the `_meta` on `a.astype(float, errors="ignore")` is potentially incorrect.

I'm not sure the best way to handle this.